### PR TITLE
fix for issue 89

### PR DESCRIPTION
**Pull request by ddbox**

### CODE REVIEW

**Comment by ddbox** [comment]
I recall we discussed a unit test for this, which I forgot about until after I made the pull request.
This failed creating a bosco submission file, it would also have been caught if we had a bosco CE to test submissions to.
I suppose we need to unit-test all supported entry types, the ones I remember off the top of my head are  bosco, htCondor, aws, gce.    I *think* slurm and pbs now sit behind htCondor job routers but I would have to dig.

File: unittests/unittest_utils.py [hunk 1]
```diff
@@ -173,5 +173,36 @@ def create_random_string(length=8):
     return "".join(random.choice(char_set) for x in range(length))
 
 
+# checks line by line that parens and quotations are balanced
+def balanced_text(myText):
```
Comment (mambelli) [inline unittests/unittest_utils.py:177]: Use a docstring for the function

PS Function is OK for this ticket. I sent an inquiry for a possible linter provided by the condor team. Maybe we ask also this Friday.

**Review by mambelli — CHANGES_REQUESTED** [review]

File: unittests/unittest_utils.py (hunk 1 above)
Comment (ddbox) [reply unittests/unittest_utils.py:177]: done

Comment (ddbox) [reply unittests/unittest_utils.py:177]: done

File: unittests/test_creation_lib_cgWParamDict.py [hunk 2]
```diff
@@ -6,15 +6,15 @@
 """
 Project:
     glideinWMS
-
+deinwms.creation.lib
```
[the remark this answers is not in the document]
Comment (ddbox) [reply unittests/test_creation_lib_cgWParamDict.py:9]: removed typo

File: unittests/fixtures/factory/config.d/Dev_Sites2.xml [hunk 3]
```diff
@@ -1,30 +1,51 @@
...
+            <attr name="GLIDEIN_Retire_Time" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="int" value="86400"/>
+            <attr name="GLIDEIN_ResourceName" const="True" glidein_publish="True" job_publish="True" parameter="True" publish="True" type="string" value="T3_US_HEPCloud_AWS"/>
+            <attr name="GLIDEIN_Site" const="True" glidein_publish="True" job_publish="True" parameter="True" publish="True" type="string" value="T3_US_HEPCloud"/>
+            <attr name="GLIDEIN_StoragePerCpuGBs" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="int" value="2"/>
+            <attr name="GLIDEIN_Supported_VOs" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="CMS,Fermilab"/>
+            <attr name="INSTANCE_TYPE" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="m3.xlarge"/>
+            <attr name="SELECTION_GROUPS" const="True" glidein_publish="False" job_publish="False" parameter="True" publish="True" type="string" value="CMS,Fermilab"/>
+
```
[the remark this answers is not in the document]
Comment (ddbox) [reply unittests/fixtures/factory/config.d/Dev_Sites2.xml:48]: done

File: unittests/fixtures/factory/config.d/Dev_Sites2.xml [hunk 4]
```diff
@@ -1,30 +1,51 @@
...
+               <per_entry glideins="1" held="1" idle="1"/>
                <per_frontends>
                </per_frontends>
             </max_jobs>
-            <release max_per_cycle="20" sleep="0.2"/>
+            <release max_per_cycle="0" sleep="0.2"/>
             <remove max_per_cycle="5" sleep="0.2"/>
-            <restrictions require_voms_proxy="False"/>
-            <submit cluster_size="10" max_per_cycle="100" sleep="0.2" slots_layout="fixed">
+            <restrictions require_glidein_glexec_use="False" require_voms_proxy="False"/>
```
[the remark this answers is not in the document]
Comment (ddbox) [reply unittests/fixtures/factory/config.d/Dev_Sites2.xml:13]: It appears you did this one

File: unittests/unittest_utils.py [hunk 5]
```diff
@@ -173,5 +173,40 @@ def create_random_string(length=8):
     return "".join(random.choice(char_set) for x in range(length))
 
 
+def balanced_text(myText):
+    """
+    checks line by line that parens and quotations are balanced
+        @type myText: string
+        @param myText: input to be checked for balanced quotations and parens
+    """
```
Comment (mambelli) [inline unittests/unittest_utils.py:181]: We use docstrings in google format:
https://cdcvs.fnal.gov/redmine/projects/glideinwms/wiki#Code-Documentation-Guidelines

PS Both VSCode and PyCharm allow to configure the docstring template to it

**Review by mambelli — CHANGES_REQUESTED** [review]

File: unittests/unittest_utils.py (hunk 5 above)
Comment (ddbox) [reply unittests/unittest_utils.py:181]: Just got this email from Steve:

The glidein submitted ran and jobs matched to it, but it went held at the end of the job for reasons I have not yet figured out.
I don't think they are tied to the bug fix but would like to hold the ticket open until I figure that out.

**Review by mambelli — APPROVED** [review]

**Pull request merged**